ws.c

```diff
@@ -0,0 +1,409 @@
+	uint16_t *u16;
+
+	if (evws->closed)
+		return;
+	evws->closed = true;
+
+	u16 = (uint16_t *)&fr[2];
+	*u16 = htons((int16_t)reason);
+	output = bufferevent_get_output(evws->bufev);
+	evbuffer_add(output, fr, 4);
+
+	/* wait for close frame writing complete and close connection */
+	bufferevent_setcb(
+		evws->bufev, NULL, close_after_write_cb, close_event_cb, evws);
+}
+
+static void
+evws_force_disconnect_(struct evws_connection *evws)
+{
+	evws_close(evws, WS_CR_NONE);
+}
+
+static enum WebSocketFrameType
+get_ws_frame(unsigned char *in_buffer, int buf_len, unsigned char **payload_ptr,
+	int *out_len)
+{
+	unsigned char opcode;
+	unsigned char fin;
+	unsigned char masked;
+	int payload_len;
+	int pos;
+	int length_field;
+	unsigned int mask;
+
+	if (buf_len < 3) {
+		return INCOMPLETE_FRAME;
+	}
+
+	opcode = in_buffer[0] & 0x0F;
+	fin = (in_buffer[0] >> 7) & 0x01;
+	masked = (in_buffer[1] >> 7) & 0x01;
+
+	payload_len = 0;
+	pos = 2;
+	length_field = in_buffer[1] & (~0x80);
+	mask = 0;
+
+	if (length_field <= 125) {
+		payload_len = length_field;
+	} else if (length_field == 126) { /* msglen is 16bit */
+		payload_len = ntohs(*(uint16_t *)(in_buffer + 2));
+		pos += 2;
+	} else if (length_field == 127) { /* msglen is 64bit */
+		payload_len = ntohs(*(uint64_t *)(in_buffer + 2));
+		pos += 8;
+	}
+	if (buf_len < payload_len + pos) {
+		return INCOMPLETE_FRAME;
+	}
+
+	if (masked) {
+		unsigned char *c;
+
+		mask = *((unsigned int *)(in_buffer + pos));
+		pos += 4;
+
+		/* unmask data */
+		c = in_buffer + pos;
+		for (int i = 0; i < payload_len; i++) {
+			c[i] = c[i] ^ ((unsigned char *)(&mask))[i % 4];
+		}
+	}
+
+	*payload_ptr = in_buffer + pos;
+	*out_len = payload_len;
+
+	switch (opcode) {
+	case 0x0:
+		return fin ? TEXT_FRAME : INCOMPLETE_TEXT_FRAME;
+	case 0x1:
+		return fin ? TEXT_FRAME : INCOMPLETE_TEXT_FRAME;
+	case 0x2:
+		return fin ? BINARY_FRAME : INCOMPLETE_BINARY_FRAME;
+	case 0x9:
+		return PING_FRAME;
+	case 0xA:
+		return PONG_FRAME;
+	default:
+		return ERROR_FRAME;
+	}
+}
+
+
+static void
+ws_evhttp_read_cb(struct bufferevent *bufev, void *arg)
+{
+	struct evws_connection *evws = arg;
+	unsigned char *payload;
+	enum WebSocketFrameType type;
+	int msg_len, in_len;
+	struct evbuffer *input = bufferevent_get_input(evws->bufev);
+
+	while ((in_len = evbuffer_get_length(input))) {
+		unsigned char *data = evbuffer_pullup(input, in_len);
+
+		type = get_ws_frame(data, in_len, &payload, &msg_len);
+
+		if (type == TEXT_FRAME) {
+			if (evws->cb) {
+				char *text = mm_malloc(msg_len + 1);
+
+				memcpy(text, payload, msg_len);
+				text[msg_len] = 0;
+				evws->cb(evws, text, msg_len, evws->cb_arg);
+				free(text);
+			}
+		} else if (type == INCOMPLETE_TEXT_FRAME || type == INCOMPLETE_FRAME) {
+			/* incomplete frame received, wait for next chunk */
+			return;
+		} else if (type == INCOMPLETE_BINARY_FRAME || type == BINARY_FRAME ||
+				   type == CLOSING_FRAME || type == ERROR_FRAME) {
+			evws_force_disconnect_(evws);
```

Comment:
So this implementation does not support binary frames. But it looks like pretty easy to implement?